doc/function.md

```diff
@@ -1,5 +1,282 @@
 # Function
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+The **Function** class provides a set of methods for creating a function object in
+native code that can later be called from JavaScript. The created function is not
+automatically visible from JavaScript. Instead it needs to be part of the add-on's
+module exports or be returned by one of the module's exported functions.
+
+In addition the `Function` class also provides methods that can be used to call
+functions that were created in JavaScript and passed to the native add-on.
+
+The `Function` class inherits its behavior from the `Object` class (for more info
+see: [Object](object.md)).
+
+## Example
+
+```cpp
+#include <napi.h>
+
+using namespace Napi;
+
+Value Fn(const CallbackInfo& info) {
+  Env env = info.Env();
+  // ...
+  return String::New(env, "Hello World");
+}
+
+Object Init(Env env, Object exports) {
+  exports.Set(String::New(env, "fn"), Function::New(env, Fn));
+}
+
+NODE_API_MODULE(NODE_GYP_MODULE_NAME, Init)
+```
+
+The above code can be used from JavaScript as follows:
+
+```js
+const addon = require('./addon');
+addon.fn();
+```
+
+With the `Function` class it is possible to call a JavaScript function object
+from a native add-on with two different methods: `Call` and `MakeCallback`.
+The API of these two methods is very similar, but they are used in different
+contexts. The `MakeCallback` method is used to call from native code back into
+JavaScript after returning from an [asynchronous operation](async_operations.md)
+and in general in situations which don't have an existing JavaScript function on
+the stack. The `Call` method is used when there is already a JavaScript function
+on the stack (for example when running a native method called from JavaScript).
+
+## Methods
+
+### Constructor
+
+Creates a new empty instance of `Function`.
+
+```cpp
+Function();
+```
+
+### Constructor
+
+Creates a new instance of the `Function` object.
+
+```cpp
+Function(napi_env env, napi_value value);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Value object.
+- `[in] value`: The `napi_value` which is a handle for a JavaScript function.
+
+Returns a non-empty `Function` instance.
```

Comment:
We should really be consistent about whether we say `Napi::Value`, `Napi::Function`, `Napi::FunctionReference`, etc. or `Value`, `Function`, or `FunctionReference`. IMO we should always add the name space in front of the class name. @mhdawson what do you think?